pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -351,6 +351,9 @@ func logResult(info *reconcileInfo, result reconcile.Result) {
 }
 
 func recordActionFailure(info *reconcileInfo, errorType metal3v1alpha1.ErrorType, errorMessage string) actionFailed {
+
+	info.host.IncrementErrorCount()
```

Comment:
If we're going to increment the counter, we need to always save the host. I think the controller only saves if we tell it the host is dirty, and this method only sets dirty based on whether the error message has a new value.

Reply:
Yes, any change to status will cause a reconciliation. We will have to watch for hot loops caused by tracking the counter value. We should only need to increment it when we transition to the relevant state, though, right? And not every time we reconcile while we are in a given state.

Does the provisioner have to count the attempts, since it knows when it is actually starting vs. continuing one?

Reply:
> Yes, any change to status will cause a reconciliation.

I'm surprised this is not kicking us into an immediate retry already (or maybe it is and we haven't noticed?).

e.g. if we have a provisioning failure now, we change the state to ProvisioningError (and maybe update the error string) and don't request a requeue. But if we get reconciled because of the state change then that should immediately bounce us back to Provisioning and try again. Is there some other magic going on?

> Does the provisioner have to count the attempts, since it knows when it is actually starting vs. continuing one?

Yeah, that might be the only way.

Reply:
> That means actionFailed.Dirty() can just always return true.
> 
What about then `handleRegistrationError()`? There `actionFailed` is used but without the dirty flag set.

Reply:
And by reviewing all the possible paths, it seems that `SetErrorMessage()` should always return true since the counter gets already increased, which leads to the original comment of @dhellmann .

Reply:
I've added a watch predicate to avoid hot loops in case of `Status.ErrorCount` update